File: Haskell/src/Modelos/Jogo.hs
{-# LANGUAGE DeriveGeneric #-}

module Modelos.Jogo (Jogo (..), JogoClass (..), inicializarJogo, carregarSave, salvarJogo, setJogador, setBot) where

import Modelos.Jogador
import Modelos.Mercado
import Modelos.Tabuleiro
import Modelos.Bot

import GHC.Generics (Generic)
import Data.Aeson (ToJSON, FromJSON, encode, decode, eitherDecode)
import qualified Data.ByteString.Lazy as B
import System.Directory (createDirectoryIfMissing)
import System.FilePath(takeDirectory)

import Data.Time (getCurrentTime, UTCTime, utcToLocalTime, LocalTime)
import Data.Time.LocalTime(hoursToTimeZone)

import System.Random(newStdGen, split)

class JogoClass jogo where
    getJogador :: jogo -> Jogador
    getBot :: jogo -> Bot
    getMercado :: jogo -> Mercado
    getDataJogo :: jogo -> LocalTime

data Jogo = Jogo {
    jogador :: Jogador,
    bot :: Bot,
    mercado :: Mercado,
    dataJogo :: LocalTime
} deriving (Show, Generic)

instance JogoClass Jogo where
    getJogador = jogador
    getBot = bot
    getMercado = mercado
    getDataJogo = dataJogo

-- Permite Conversão para JSON

instance ToJSON Jogo
instance FromJSON Jogo

-- Função para mudar o Jogo quando houver alteração no Jogador

setJogador :: Jogador -> Jogo -> Jogo
setJogador novoJogador jogo = jogo { jogador = novoJogador }

setBot :: Bot -> Jogo -> Jogo
setBot novoBot jogo = jogo { bot = novoBot }

-- Função para salvar e carregar o jogo com um JSON
-- O FilePath tem o formato "src/BD/save{numero}.json"

salvarJogo :: FilePath -> Jogo -> IO ()
salvarJogo caminho jogo = do
    let diretorio = takeDirectory caminho
    createDirectoryIfMissing True diretorio  
    utcTime <- getCurrentTime
    let zonaBrasilia = hoursToTimeZone (-3)
    let agora = utcToLocalTime zonaBrasilia utcTime
    let jogoAtualizado = jogo { dataJogo = agora }
    B.writeFile caminho (encode jogoAtualizado)

carregarSave :: FilePath -> IO (Maybe Jogo)
carregarSave caminho = do
    conteudo <- B.readFile caminho
    return (decode conteudo)

-- Função de inicialização

inicializarJogo :: String -> IO Jogo
inicializarJogo nomeJogador = do
    utcTime <- getCurrentTime
    let zonaBrasilia = hoursToTimeZone (-3)  
    let agora = utcToLocalTime zonaBrasilia utcTime

    gen <- newStdGen
    let (genJogador, genBot) = split gen

    let tabelaJogador = geraTabela genJogador
    let tabelaBot = geraTabela genBot

    return $ Jogo (iniciarJogador nomeJogador tabelaJogador) (iniciarBot tabelaBot) iniciarMercado agora

-- Funções auxiliares de inicialização

iniciarJogador :: String -> Tabela -> Jogador
iniciarJogador nomeJogador tabelaJogador = Jogador nomeJogador 0 tabelaJogador 30 0 0 0

iniciarBot :: Tabela -> Bot
iniciarBot tabelaBot = Bot tabelaBot 30 3 1 0

iniciarMercado :: Mercado
iniciarMercado = Mercado 250 400 350

File: Prolog/src/Modelos/Jogo.pl
:- module(jogo, [
    inicializarJogo/2, carregarSave/2, salvarJogo/2, 
    setJogador/3, setBot/3, 
    getJogador/2, getBot/2, 
    getMercado/2, getDataJogo/2
]).

:- use_module('./Jogador').
:- use_module('./Mercado').
:- use_module('./Tabuleiro').
:- use_module('./Bot.pl').
:- use_module(library(time)).  
:- use_module(library(filesex)).

:- dynamic jogo/4.

% jogo(Jogador, Bot, Mercado, DataJogo) // Em teoria, falta o Bot.

% Getters

getJogador(jogo(Jogador, _, _, _), Jogador).
getBot(jogo(_, Bot, _, _), Bot).
getMercado(jogo(_, _, Mercado, _), Mercado).
getDataJogo(jogo(_, _, _, DataJogo), DataJogo).

% Setters

setJogador(jogo(_, Bot, Mercado, DataJogo), NovoJogador, jogo(NovoJogador, Bot, Mercado, DataJogo)).
setBot(jogo(Jogador, _, Mercado, DataJogo), NovoBot, jogo(Jogador, NovoBot, Mercado, DataJogo)).

% Salvar jogo em arquivo (../BD/save{num})

salvarJogo(Caminho, Jogo) :-
    open(Caminho, write, Stream),
    write_canonical(Stream, Jogo), 
    write(Stream, '.\n'),  
    close(Stream).

% Carregar jogo de arquivo

carregarSave(Caminho, Jogo) :-
    ( exists_file(Caminho) ->
        open(Caminho, read, Stream),
        read(Stream, Jogo),  
        close(Stream)
    ;   fail 
    ).

% Inicialização do jogo
% Tirar o _ ao fazer o Bot e substituir por ele

inicializarJogo(NomeJogador, jogo(Jogador, Bot, Mercado, DataJogo)) :-
    get_time(TempoAtual),
    format_time(atom(DataJogo), '%Y-%m-%dT%H:%M:%S', TempoAtual),
    geraTabuleiroDispor(TabelaJogador),
    geraTabuleiroDispor(TabelaBot),
    iniciarJogador(NomeJogador, TabelaJogador, Jogador),
    iniciarBot(TabelaBot, Bot),
    Mercado = mercado(250, 400, 350).

% Inicializar Jogador

iniciarJogador(Nome, Tabela, jogador(Nome, 0, Tabela, 55, 2, 1, 0)).

% Inicializar Bot

iniciarBot(Tabela, Bot) :- getDefaultBot(Tabela, Bot).